cardano-cli/test/cardano-cli-test-lib/Test/Cardano/CLI/Util.hs

```diff
@@ -284,7 +290,32 @@ redactJsonField fieldName replacement sourceFilePath targetFilePath = GHC.withFr
         v -> pure v
       H.evalIO $ LBS.writeFile targetFilePath (Aeson.encodePretty redactedJson)
 
--- | Run action acquiring a semaphore, and releasing afterwards. Allows to guard against concurrent access to
--- a block of code
-bracketSem :: MonadBaseControl IO m => QSem -> m c -> m c
-bracketSem semaphore = bracket_ (liftBase $ waitQSem semaphore) (liftBase $ signalQSem semaphore)
+
+-- | A file semaphore protecting against a concurrent path access
+data FileSem = FileSem !FilePath !QSem
+
+instance Show FileSem where
+  show (FileSem path _) = "FileSem " ++ path
+
+deriving via (ShowOf FileSem) instance Pretty FileSem
+
+-- | Create new file semaphore. Always use with @NOINLINE@ pragma! Example:
```

Comment:
Otherwise GHC could inline the semaphore, and instead of one we could end up with multiple, different semaphores not communicating with each other. So the locking wouldn't work.